receiver/kafkametricsreceiver/topic_scraper.go

```diff
@@ -0,0 +1,120 @@
+// Copyright  OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kafkametricsreceiver
+
+import (
+	"context"
+	"regexp"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/consumer/simple"
+	"go.opentelemetry.io/collector/receiver/scraperhelper"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkametricsreceiver/internal/metadata"
+)
+
+type topicScraper struct {
+	client       sarama.Client
+	logger       *zap.Logger
+	topicFilter  *regexp.Regexp
+	saramaConfig *sarama.Config
+	config       Config
+}
+
+func (s *topicScraper) Name() string {
+	return "topics"
+}
+
+func (s *topicScraper) shutdown(context.Context) error {
+	if !s.client.Closed() {
+		return s.client.Close()
+	}
+	return nil
+}
+
+func (s *topicScraper) scrape(context.Context) (pdata.ResourceMetricsSlice, error) {
+	topics, err := s.client.Topics()
+	metrics := simple.Metrics{
+		Metrics:                    pdata.NewMetrics(),
+		MetricFactoriesByName:      metadata.M.FactoriesByName(),
+		InstrumentationLibraryName: InstrumentationLibName,
+		Timestamp:                  time.Now(),
+	}
+	if err != nil {
+		return metrics.Metrics.ResourceMetrics(), err
+	}
+
+	var matchedTopics []string
+	for _, t := range topics {
+		if s.topicFilter.MatchString(t) {
+			matchedTopics = append(matchedTopics, t)
+		}
+	}
+	for _, topic := range matchedTopics {
+		partitions, err := s.client.Partitions(topic)
+		if err != nil {
+			continue
+		}
+		topicMetrics := metrics.WithLabels(map[string]string{
+			metadata.L.Topic: topic,
+		})
+		topicMetrics.AddGaugeDataPoint(metadata.M.KafkaTopicPartitions.Name(), int64(len(partitions)))
+		for _, partition := range partitions {
+			partitionMetrics := topicMetrics.WithLabels(map[string]string{
+				metadata.L.Partition: string(partition),
+			})
+			currentOffset, err := s.client.GetOffset(topic, partition, sarama.OffsetNewest)
+			if err == nil {
```

Comment:
What about all these? Are they expected to fail often? Of course the issue is if something is continuously failing it'll log every scrape interval, but it may be hard to diagnose missing metrics when errors aren't logged.

Reply:
These are not expected to fail but in the case of an error, at the next scrape interval, it will be caught and logged at #L59.
Logging here would be very chatty. Its essentially no. of partitions * no. of topics * interval.